File: open-policy-platform/services/feature-flags/sdk/python/feature_flags.py
"""
Python SDK for OpenPolicy Platform Feature Flags
"""

import os
import json
import logging
from typing import Any, Dict, List, Optional, Union
from dataclasses import dataclass, asdict
import requests
from functools import lru_cache
import threading
import time

logger = logging.getLogger(__name__)

@dataclass
class FlagContext:
    """Context for evaluating feature flags"""
    user_id: Optional[str] = None
    user_email: Optional[str] = None
    user_role: Optional[str] = None
    organization_id: Optional[str] = None
    environment: str = "production"
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None
    custom_attributes: Dict[str, Any] = None
    
    def __post_init__(self):
        if self.custom_attributes is None:
            self.custom_attributes = {}

class FeatureFlagClient:
    """Client for feature flag service"""
    
    def __init__(
        self,
        service_url: str = None,
        api_key: Optional[str] = None,
        default_timeout: int = 5,
        enable_cache: bool = True,
        cache_ttl: int = 300,
        fallback_values: Optional[Dict[str, Any]] = None
    ):
        self.service_url = service_url or os.getenv("FEATURE_FLAG_SERVICE_URL", "http://localhost:9024")
        self.api_key = api_key or os.getenv("FEATURE_FLAG_API_KEY")
        self.timeout = default_timeout
        self.enable_cache = enable_cache
        self.cache_ttl = cache_ttl
        self.fallback_values = fallback_values or {}
        self._cache = {}
        self._cache_timestamps = {}
        self._lock = threading.Lock()
        
        # Session for connection pooling
        self.session = requests.Session()
        if self.api_key:
            self.session.headers["X-API-Key"] = self.api_key
    
    def _is_cache_valid(self, key: str) -> bool:
        """Check if cached value is still valid"""
        if key not in self._cache_timestamps:
            return False
        return time.time() - self._cache_timestamps[key] < self.cache_ttl
    
    def _get_from_cache(self, key: str) -> Optional[Any]:
        """Get value from cache if valid"""
        with self._lock:
            if self.enable_cache and self._is_cache_valid(key):
                return self._cache.get(key)
        return None
    
    def _set_cache(self, key: str, value: Any):
        """Set value in cache"""
        with self._lock:
            self._cache[key] = value
            self._cache_timestamps[key] = time.time()
    
    def _make_request(self, method: str, endpoint: str, **kwargs) -> Optional[Dict[str, Any]]:
        """Make HTTP request to feature flag service"""
        url = f"{self.service_url}{endpoint}"
        kwargs.setdefault("timeout", self.timeout)
        
        try:
            response = self.session.request(method, url, **kwargs)
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            logger.error(f"Feature flag request failed: {e}")
            return None
    
    def evaluate(
        self,
        flag_key: str,
        context: Union[FlagContext, Dict[str, Any]],
        default_value: Any = None
    ) -> Any:
        """Evaluate a single feature flag"""
        # Convert dict to FlagContext if needed
        if isinstance(context, dict):
            context = FlagContext(**context)
        
        # Generate cache key
        cache_key = f"{flag_key}:{hash(json.dumps(asdict(context), sort_keys=True))}"
        
        # Check cache
        cached_value = self._get_from_cache(cache_key)
        if cached_value is not None:
            return cached_value
        
        # Make request
        response = self._make_request(
            "POST",
            f"/evaluate/{flag_key}",
            json=asdict(context)
        )
        
        if response:
            value = response.get("value", default_value)
            self._set_cache(cache_key, value)
            return value
        
        # Fall back to configured fallback or provided default
        return self.fallback_values.get(flag_key, default_value)
    
    def evaluate_batch(
        self,
        flag_keys: List[str],
        context: Union[FlagContext, Dict[str, Any]],
        default_values: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """Evaluate multiple feature flags"""
        # Convert dict to FlagContext if needed
        if isinstance(context, dict):
            context = FlagContext(**context)
        
        default_values = default_values or {}
        
        # Check cache for all flags
        results = {}
        uncached_keys = []
        
        for flag_key in flag_keys:
            cache_key = f"{flag_key}:{hash(json.dumps(asdict(context), sort_keys=True))}"
            cached_value = self._get_from_cache(cache_key)
            if cached_value is not None:
                results[flag_key] = cached_value
            else:
                uncached_keys.append(flag_key)
        
        # Fetch uncached flags
        if uncached_keys:
            response = self._make_request(
                "POST",
                "/evaluate/batch",
                json={
                    "flag_keys": uncached_keys,
                    "context": asdict(context)
                }
            )
            
            if response and "evaluations" in response:
                for flag_key, evaluation in response["evaluations"].items():
                    value = evaluation.get("value")
                    if value is not None:
                        results[flag_key] = value
                        # Cache the result
                        cache_key = f"{flag_key}:{hash(json.dumps(asdict(context), sort_keys=True))}"
                        self._set_cache(cache_key, value)
        
        # Apply defaults for missing flags
        for flag_key in flag_keys:
            if flag_key not in results:
                results[flag_key] = default_values.get(
                    flag_key,
                    self.fallback_values.get(flag_key)
                )
        
        return results
    
    def get_all_flags(self) -> Optional[List[Dict[str, Any]]]:
        """Get all feature flags"""
        response = self._make_request("GET", "/flags")
        if response:
            return response.get("flags", [])
        return None
    
    def clear_cache(self):
        """Clear local cache"""
        with self._lock:
            self._cache.clear()
            self._cache_timestamps.clear()

# Decorator for feature flags
class feature_flag:
    """Decorator for feature flag controlled functions"""
    
    def __init__(
        self,
        flag_key: str,
        default_value: bool = False,
        client: Optional[FeatureFlagClient] = None
    ):
        self.flag_key = flag_key
        self.default_value = default_value
        self.client = client or _default_client
    
    def __call__(self, func):
        def wrapper(*args, **kwargs):
            # Extract context from function arguments
            context = kwargs.get("flag_context") or FlagContext()
            
            # Evaluate flag
            if self.client.evaluate(self.flag_key, context, self.default_value):
                return func(*args, **kwargs)
            else:
                # Return None or raise exception based on function
                return None
        
        return wrapper

# Global default client
_default_client = FeatureFlagClient()

# Convenience functions
def init(
    service_url: str = None,
    api_key: str = None,
    **kwargs
):
    """Initialize the default feature flag client"""
    global _default_client
    _default_client = FeatureFlagClient(
        service_url=service_url,
        api_key=api_key,
        **kwargs
    )

def evaluate(flag_key: str, context: Union[FlagContext, Dict[str, Any]], default_value: Any = None) -> Any:
    """Evaluate a feature flag using the default client"""
    return _default_client.evaluate(flag_key, context, default_value)

def evaluate_batch(
    flag_keys: List[str],
    context: Union[FlagContext, Dict[str, Any]],
    default_values: Optional[Dict[str, Any]] = None
) -> Dict[str, Any]:
    """Evaluate multiple feature flags using the default client"""
    return _default_client.evaluate_batch(flag_keys, context, default_values)

# Example usage
if __name__ == "__main__":
    # Initialize client
    client = FeatureFlagClient(
        service_url="http://localhost:9024",
        fallback_values={
            "new_ui": False,
            "dark_mode": True,
            "api_v2": False
        }
    )
    
    # Create context
    context = FlagContext(
        user_id="user-123",
        user_email="user@example.com",
        user_role="admin",
        organization_id="org-456",
        custom_attributes={
            "plan": "premium",
            "country": "US"
        }
    )
    
    # Evaluate single flag
    new_ui_enabled = client.evaluate("new_ui", context, default_value=False)
    print(f"New UI enabled: {new_ui_enabled}")
    
    # Evaluate multiple flags
    flags = client.evaluate_batch(
        ["new_ui", "dark_mode", "api_v2"],
        context
    )
    print(f"Flags: {flags}")
    
    # Using decorator
    @feature_flag("experimental_feature", default_value=False)
    def experimental_function():
        return "This is experimental!"
    
    result = experimental_function(flag_context=context)
    print(f"Experimental result: {result}")

File: open-policy-platform/services/feature-flags/sdk/javascript/feature-flags.ts
/**
 * TypeScript SDK for OpenPolicy Platform Feature Flags
 */

export interface FlagContext {
  userId?: string;
  userEmail?: string;
  userRole?: string;
  organizationId?: string;
  environment?: string;
  ipAddress?: string;
  userAgent?: string;
  customAttributes?: Record<string, any>;
}

export interface FlagEvaluation {
  key: string;
  value: any;
  variationIndex?: number;
  reason: string;
  source: 'local' | 'launchdarkly';
}

export interface FeatureFlagClientConfig {
  serviceUrl?: string;
  apiKey?: string;
  defaultTimeout?: number;
  enableCache?: boolean;
  cacheTTL?: number;
  fallbackValues?: Record<string, any>;
  enableLogging?: boolean;
}

class Cache {
  private cache: Map<string, { value: any; timestamp: number }> = new Map();
  private ttl: number;

  constructor(ttl: number = 300000) { // 5 minutes default
    this.ttl = ttl;
  }

  set(key: string, value: any): void {
    this.cache.set(key, {
      value,
      timestamp: Date.now()
    });
  }

  get(key: string): any | undefined {
    const item = this.cache.get(key);
    if (!item) return undefined;

    if (Date.now() - item.timestamp > this.ttl) {
      this.cache.delete(key);
      return undefined;
    }

    return item.value;
  }

  clear(): void {
    this.cache.clear();
  }
}

export class FeatureFlagClient {
  private serviceUrl: string;
  private apiKey?: string;
  private timeout: number;
  private enableCache: boolean;
  private cache: Cache;
  private fallbackValues: Record<string, any>;
  private enableLogging: boolean;

  constructor(config: FeatureFlagClientConfig = {}) {
    this.serviceUrl = config.serviceUrl || process.env.FEATURE_FLAG_SERVICE_URL || 'http://localhost:9024';
    this.apiKey = config.apiKey || process.env.FEATURE_FLAG_API_KEY;
    this.timeout = config.defaultTimeout || 5000;
    this.enableCache = config.enableCache !== false;
    this.cache = new Cache(config.cacheTTL);
    this.fallbackValues = config.fallbackValues || {};
    this.enableLogging = config.enableLogging || false;
  }

  private log(message: string, data?: any): void {
    if (this.enableLogging) {
      console.log(`[FeatureFlags] ${message}`, data || '');
    }
  }

  private generateCacheKey(flagKey: string, context: FlagContext): string {
    const contextStr = JSON.stringify(context, Object.keys(context).sort());
    return `${flagKey}:${this.hashCode(contextStr)}`;
  }

  private hashCode(str: string): number {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
      const char = str.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash; // Convert to 32bit integer
    }
    return hash;
  }

  private async makeRequest<T>(
    method: string,
    endpoint: string,
    data?: any
  ): Promise<T | null> {
    const url = `${this.serviceUrl}${endpoint}`;
    const headers: HeadersInit = {
      'Content-Type': 'application/json',
    };

    if (this.apiKey) {
      headers['X-API-Key'] = this.apiKey;
    }

    try {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), this.timeout);

      const response = await fetch(url, {
        method,
        headers,
        body: data ? JSON.stringify(data) : undefined,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      this.log('Request failed', { endpoint, error });
      return null;
    }
  }

  async evaluate(
    flagKey: string,
    context: FlagContext,
    defaultValue?: any
  ): Promise<any> {
    // Check cache first
    if (this.enableCache) {
      const cacheKey = this.generateCacheKey(flagKey, context);
      const cachedValue = this.cache.get(cacheKey);
      if (cachedValue !== undefined) {
        this.log('Cache hit', { flagKey, value: cachedValue });
        return cachedValue;
      }
    }

    // Make request
    const response = await this.makeRequest<FlagEvaluation>(
      'POST',
      `/evaluate/${flagKey}`,
      context
    );

    if (response) {
      const value = response.value;
      
      // Cache the result
      if (this.enableCache) {
        const cacheKey = this.generateCacheKey(flagKey, context);
        this.cache.set(cacheKey, value);
      }

      this.log('Evaluated flag', { flagKey, value });
      return value;
    }

    // Fall back to configured fallback or provided default
    const fallbackValue = this.fallbackValues[flagKey] ?? defaultValue;
    this.log('Using fallback value', { flagKey, value: fallbackValue });
    return fallbackValue;
  }

  async evaluateBatch(
    flagKeys: string[],
    context: FlagContext,
    defaultValues?: Record<string, any>
  ): Promise<Record<string, any>> {
    const results: Record<string, any> = {};
    const uncachedKeys: string[] = [];

    // Check cache for all flags
    if (this.enableCache) {
      for (const flagKey of flagKeys) {
        const cacheKey = this.generateCacheKey(flagKey, context);
        const cachedValue = this.cache.get(cacheKey);
        if (cachedValue !== undefined) {
          results[flagKey] = cachedValue;
        } else {
          uncachedKeys.push(flagKey);
        }
      }
    } else {
      uncachedKeys.push(...flagKeys);
    }

    // Fetch uncached flags
    if (uncachedKeys.length > 0) {
      const response = await this.makeRequest<{ evaluations: Record<string, FlagEvaluation> }>(
        'POST',
        '/evaluate/batch',
        {
          flag_keys: uncachedKeys,
          context
        }
      );

      if (response?.evaluations) {
        for (const [flagKey, evaluation] of Object.entries(response.evaluations)) {
          const value = evaluation.value;
          results[flagKey] = value;

          // Cache the result
          if (this.enableCache && value !== null) {
            const cacheKey = this.generateCacheKey(flagKey, context);
            this.cache.set(cacheKey, value);
          }
        }
      }
    }

    // Apply defaults for missing flags
    for (const flagKey of flagKeys) {
      if (!(flagKey in results)) {
        results[flagKey] = defaultValues?.[flagKey] ?? this.fallbackValues[flagKey];
      }
    }

    this.log('Batch evaluation complete', { flagKeys, results });
    return results;
  }

  async getAllFlags(): Promise<any[] | null> {
    const response = await this.makeRequest<{ flags: any[] }>('GET', '/flags');
    return response?.flags || null;
  }

  clearCache(): void {
    this.cache.clear();
    this.log('Cache cleared');
  }
}

// React hooks
import { useState, useEffect, useContext, createContext, ReactNode } from 'react';

interface FeatureFlagProviderProps {
  client: FeatureFlagClient;
  context: FlagContext;
  children: ReactNode;
}

interface FeatureFlagContextValue {
  client: FeatureFlagClient;
  context: FlagContext;
  flags: Record<string, any>;
  loading: boolean;
  error: Error | null;
  refetch: () => Promise<void>;
}

const FeatureFlagContext = createContext<FeatureFlagContextValue | null>(null);

export function FeatureFlagProvider({
  client,
  context,
  children
}: FeatureFlagProviderProps) {
  const [flags, setFlags] = useState<Record<string, any>>({});
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<Error | null>(null);

  const fetchFlags = async () => {
    try {
      setLoading(true);
      setError(null);
      
      // Get all flags first
      const allFlags = await client.getAllFlags();
      if (allFlags) {
        const flagKeys = allFlags.map(f => f.key);
        const evaluations = await client.evaluateBatch(flagKeys, context);
        setFlags(evaluations);
      }
    } catch (err) {
      setError(err as Error);
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    fetchFlags();
  }, [context.userId, context.organizationId]); // Re-fetch when key context changes

  const value: FeatureFlagContextValue = {
    client,
    context,
    flags,
    loading,
    error,
    refetch: fetchFlags
  };

  return (
    <FeatureFlagContext.Provider value={value}>
      {children}
    </FeatureFlagContext.Provider>
  );
}

export function useFeatureFlag(flagKey: string, defaultValue?: any): any {
  const context = useContext(FeatureFlagContext);
  
  if (!context) {
    throw new Error('useFeatureFlag must be used within a FeatureFlagProvider');
  }

  const [value, setValue] = useState(context.flags[flagKey] ?? defaultValue);

  useEffect(() => {
    // If flag is not in the pre-fetched flags, fetch it individually
    if (!(flagKey in context.flags) && !context.loading) {
      context.client.evaluate(flagKey, context.context, defaultValue)
        .then(setValue);
    } else {
      setValue(context.flags[flagKey] ?? defaultValue);
    }
  }, [flagKey, context.flags, context.loading]);

  return value;
}

export function useFeatureFlags(flagKeys: string[], defaultValues?: Record<string, any>): Record<string, any> {
  const context = useContext(FeatureFlagContext);
  
  if (!context) {
    throw new Error('useFeatureFlags must be used within a FeatureFlagProvider');
  }

  const [values, setValues] = useState<Record<string, any>>({});

  useEffect(() => {
    const missingKeys = flagKeys.filter(key => !(key in context.flags));
    
    if (missingKeys.length > 0 && !context.loading) {
      context.client.evaluateBatch(missingKeys, context.context, defaultValues)
        .then(newValues => {
          setValues(prev => ({ ...prev, ...newValues }));
        });
    }

    // Set values for flags we already have
    const currentValues: Record<string, any> = {};
    for (const key of flagKeys) {
      currentValues[key] = context.flags[key] ?? defaultValues?.[key];
    }
    setValues(currentValues);
  }, [flagKeys.join(','), context.flags, context.loading]);

  return values;
}

// HOC for feature flags
export function withFeatureFlag<P extends object>(
  flagKey: string,
  defaultValue: boolean = false
) {
  return function (Component: React.ComponentType<P>) {
    return function WithFeatureFlagComponent(props: P) {
      const enabled = useFeatureFlag(flagKey, defaultValue);
      
      if (!enabled) {
        return null;
      }

      return <Component {...props} />;
    };
  };
}

// Default client instance
let defaultClient: FeatureFlagClient | null = null;

export function initFeatureFlags(config: FeatureFlagClientConfig): FeatureFlagClient {
  defaultClient = new FeatureFlagClient(config);
  return defaultClient;
}

export function getDefaultClient(): FeatureFlagClient {
  if (!defaultClient) {
    defaultClient = new FeatureFlagClient();
  }
  return defaultClient;
}